drakrun/drakrun/storage.py

```diff
@@ -308,9 +310,247 @@ def import_vm0(self, path: str):
+        try:
+            logging.info("Creating new volume vm-0")
+            subprocess.run(
+                [
+                    "lvcreate",
+                    "-L", disk_size,
+                    "-n", "vm-0",
+                    self.lvm_volume_group
+                ],
+                check=True
+            )
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to create a new volume using lvcreate.")
+
+    def snapshot_vm0_volume(self):
+        """ Saves or snapshots base vm-0 volume for later use by other VMs """
+        try:
+            logging.info("Creating snapshot of vm-0")
+            subprocess.check_output(
+                [
+                    "lvcreate",
+                    "-s",
+                    "-L", self.install_info.disk_size,
+                    "-n", "vm-0-snap",
+                    f"{self.lvm_volume_group}/vm-0"
+                ],
+                stderr=subprocess.STDOUT,
+            )
+        except subprocess.CalledProcessError as exc:
+            if b"already exists in volume group" not in exc.output:
+                raise RuntimeError("Failed to create a snapshot of vm-0")
+
+    def get_vm_disk_path(self, vm_id: int) -> str:
+        """ Returns disk path for given VM as defined by XL configuration """
+        return f"phy:/dev/{self.lvm_volume_group}/vm-{vm_id},hda,w"
+
+    def rollback_vm_storage(self, vm_id: int):
+        """ Rolls back changes and prepares fresh storage for new run of this VM """
+        vm_id_vol_snap = os.path.join("/dev", f"{self.lvm_volume_group}", f"vm-{vm_id}-snap")
+
+        # Checking if snapshot exists as that is the last step
+        # which will imply that the disk already exists
+        # if rollback required
+        if not os.path.exists(vm_id_vol_snap):
+            try:
+                logging.info(f"Creating vm-{vm_id}")
+                subprocess.check_output(
+                    ' '.join(
+                        [
+                            "lvcreate",
+                            "-L", self.install_info.disk_size,
+                            "-n", f"vm-{vm_id}",
+                            self.lvm_volume_group
+                        ]
+                    ),
+                    shell=True,
+                )
+
+                logging.info(f"Cloning data to vm-{vm_id} from vm-0 snap")
+
+                # couldn't find anything like zfs clone, any better ideas for performance improvements?
+                subprocess.check_output(
+                    ' '.join(
+                        [
+                            "dd",
+                            f"if=/dev/{self.lvm_volume_group}/vm-0-snap",
+                            f"of=/dev/{self.lvm_volume_group}/vm-{vm_id}",
+                            "status=progress"
+                        ]
+                    ),
+                    shell=True,
+                )
```

Comment:
Did I understand the working of the function wrong? Looking at the other storage backends, it seemed like rollback meant rolling back the `vm_id` disk to `vm_0` snapshot disks so that we can just reset the state for new VMs and then later discard them and rollback to `vm_0` state. Am I wrong here?